src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "seastarx.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/sharded.hh>
+
+#include <boost/lockfree/queue.hpp>
+
+#include <atomic>
+#include <condition_variable>
+#include <memory>
+#include <semaphore.h>
+
+namespace v8_engine {
+
+namespace internal {
+
+// This class is used for notify seastar thread when executor thread will
+// complete task
+class condition {
+public:
+    condition();
+
+    // Run this function in seastar thread.
+    ss::future<> wait();
+
+    // Run this function in executor thread
+    void notify();
+
+private:
+    ss::file_desc _file_desc;
+    int _fd;
+    ss::pollable_fd _pollable_fd;
+    eventfd_t _event{0};
+};
+
+// Abstract class for executor task
+struct work_item {
+    virtual ~work_item() {}
+    virtual void process() = 0;
+};
+
+// This class implement task for executor. Contains promise logic
+template<typename Func>
+struct task final : work_item {
+    explicit task(Func&& f)
+      : _func(std::move(f)) {}
+
+    // Process task from executor in executor thread and set state in ss::future
+    void process() override {
+        try {
+            _func();
+            _state.set(true);
+        } catch (...) {
+            _state.set_exception(std::current_exception());
+        }
+        _on_done.notify();
+    }
+
+    // Get future for seastar thread for wating when executor thread will
+    // complete task
+    ss::future<> get_future() {
+        return _on_done.wait().then([this] {
+            if (_state.failed()) {
+                return ss::make_exception_future<>(
+                  std::move(_state).get_exception());
+            } else {
+                return ss::make_ready_future<>();
+            }
+        });
+    }
+
+    Func _func;
+    ss::future_state<bool> _state;
```

Comment:
`process` == `p.set_value()`
`get_future` == `p.get_future()`
The only difference left would be to manually call `_on_done.notify()`

Reply:
Got this I need to use std::promise
For waiting result I have to use wait_for for std::future but `This function may block for longer than timeout_duration due to scheduling or resource contention delays. ` and it can steal reactor time

wait will be smth like that
```
do {
        status = future.wait_for(std::chrono::miliseconds(0));
       if (status == std::future_status::ready) {
            return ss::now();
        }
    } while (status != std::future_status::ready); 
```

Reply:
ah right , well that is fine ! I like this !

Reply:
I'm wondering, why `ss::future_stae<bool>` contains bool since the only value which is set there is true? Also, there is no condition that checks the value.
Also, maybe the simple `std::exception_ptr` will work in this case. Inside the `get_future` you can wait on `_on_done` and when it's activated you can check if the exception_ptr is set. If it's set you can return exception future and OK future can be returned otherwise.

Reply:
Done